vinyl: fix read stalled by write

locker opened the pull request:
The Vinyl read iterator, which is used for serving range select requests, works as follows:
1. Scan in-memory sources. If found an exact match or a chain in the cache, return it.
2. If not found, scan disk sources. This operation may yield.
3. If any new data was inserted into the active memory tree, go to step 1, effectively restarting the iteration from the same key.

Apparently, such an algorithm doesn't guarantee any progress of a read operation at all - when we yield reading disk on step 2 after a restart, even newer data may be inserted into the active memory tree, forcing us to restart again. In other words, in presence of an intensive write workload, read ops rate may drop down to literally 0.

It hasn't always been like so. Before commit 83462a5cf163, we only restored the memory tree iterator after a yield, without restarting the whole procedure. This makes sense, because only memory tree may change after a yield so there's no point in rescanning other sources, including disk. By restarting iteration after a yield, the above-mentioned commit fixed bug #3395: initially we assumed that statements may never be deleted from a memory tree while actually they can be deleted by rollback after a failed WAL write.

This PR reverts the above-mentioned commit and re-fixes bug #3395 in a proper way.

Closes #5700

Gerold103 reviewed (approved):
Thanks for the patch! Looks good given that I don't remember vinyl too good. But I have one question - why even bother with these re-evaluations? Why can't just proceed with returning next data? Worst case - you won't notice some latest changes, and so what? It would happen anyway if the changes would be made a little later. I mean, in vinyl select you can't guarantee when it yields, which data it would notice in time, which it would miss because it walked past a key too fast and the key would be updated shortly after.

locker commented:
> But I have one question - why even bother with these re-evaluations? Why can't just proceed with returning next data? Worst case - you won't notice some latest changes, and so what? It would happen anyway if the changes would be made a little later. I mean, in vinyl select you can't guarantee when it yields, which data it would notice in time, which it would miss because it walked past a key too fast and the key would be updated shortly after.

Right after we read a tuple, we add it to the transaction read set so if a tuple is updated shortly after a read iterator returned it, the read iterator will be sent to read view or aborted. This guarantees serializability. However, if a read iterator returned a tuple that was already deleted from the memory tree by rollback on WAL error, it wouldn't be sent to read view or aborted, breaking transaction serializability (returning phantom tuples).

Another reason why a read iterator must always return latest tuples unless it's in read view, because we may add them to the cache, and the cache is supposed to store only the latest data.

So if a tuple was deleted from the memory level while we yielded, we need to check if it was the min tuple that was deleted (the one that we were going to return to the user). If we deleted another tuple, we don't care. If we deleted the min tuple, we need to choose the next min tuple. So I "reevaluate" sources, i.e. iterate over all sources and choose the next min tuple.

locker commented:
Cherry-picked to 2.10 and 1.10.
